ranger/config/commands.py

```diff
@@ -1176,24 +1177,86 @@ def execute(self):
 
 
 class chmod(Command):
-    """:chmod <octal number>
+    """:chmod [<octal number>|<mode-spec>]
 
-    Sets the permissions of the selection to the octal number.
+    Sets the permissions of the selection to the given mode.
 
     The octal number is between 0 and 777. The digits specify the
     permissions for the user, the group and others.
 
     A 1 permits execution, a 2 permits writing, a 4 permits reading.
     Add those numbers to combine them. So a 7 permits everything.
+
+    The <mode-spec> is a comma-separated where each entry has the
+    form [ugoa]+[+-][rwx]+
     """
 
-    def execute(self):
-        mode_str = self.rest(1)
+    def __init__(self, *args, **kwargs):
+        '''init chmod-command.'''
+        super(chmod, self).__init__(*args, **kwargs)
+        self.re_mode = None
+        self.bitmasks = None
+
+    def prepare(self):
```

Comment:
This seems like it fits more in `__init__` rather than `execute`. Whether or not it's still in its own function is up to you.

Reply:
I moved it out of `__init__` because of ranger's behaviour:  if somebody types `:`, `c`, `h`, `m`, etc. at ranger's command line then `__init__` will be called at every keystroke (when there is only one matching command; in the `Console.on_line_change` for the "quick-Test/Call"). That's why I didn't want to stuff much initialization in `__init__` in order to not slow down this typing thing. After finding out that there is no `quick` the whole instance is thrown away.

Reply:
👍